### PR TITLE
Framework: Add support for start threshold

### PR DESCRIPTION
**Pull request by Guldoman**
With newer kernels, Framework laptops support setting both start and end thresholds.

Without this change only the end threshold is set, so the laptop never starts charging.

Note: you need to disable all other battery threshold mechanisms (bios, ectool, ...) and set `probe_with_fwk_charge_control=Y` in the `cros_charge_control` kernel module.

### CODE REVIEW

File: devices/Framework.js [hunk 1]
```diff
@@ -19,7 +20,7 @@ export const FrameworkSingleBatteryBAT1 = GObject.registerClass({
         this.type = 31;
         this.deviceNeedRootPermission = true;
         this.deviceHaveDualBattery = false;
-        this.deviceHaveStartThreshold = false;
+        this.deviceHaveStartThreshold = fileExists(BAT1_START_PATH);
```
Comment (Guldoman) [inline devices/Framework.js]: Is this problematic here?
Should it be moved in `isAvaliable`?

**Comment by maniacx** [comment]
I will detect. But however, it will still write only end threshold.
Since framework_tool doesnt support this yet and  logic will get quite complex, since we still have to support laptops with end_threshold only, I would make another class instead, this way we do not have to complicate thing for laptops/kernel using only end_threshold.

I do not have the time currently, that is why all my extensions are on hold, for at least 3 weeks. But you can do it and test it out. 

Currently in Framework.js we have only 1 class.
**FrameworkSingleBatteryBAT1**

Make another one class for device /kernel running start and end threshold

Code will be the same as class **System76SingleBattery** found here https://github.com/maniacx/Battery-Health-Charging/blob/GNOME45/devices/System76.js

Copy paste the class, rename it to another one. example
**FrameworkSingleBatteryEndStartBAT1**
Rename the vendor to 
```
        if (!dmiVendor()?.includes('Framework'))
            return false;
```
change the type to 
```
        this.type = 36;
```

Change also to a different one
```
        this.name = 'FrameworkEndStartLimit';
```

Add the class to device list

https://github.com/maniacx/Battery-Health-Charging/blob/GNOME45/lib/deviceList.js
```
    Framework.FrameworkSingleBatteryEndStartBAT1,             // 36
```
I think that about it. You should see start end options in quick settings and in preferences.

**Comment by maniacx** [comment]
I forgot. You will also need to add something to force reject if start_threshold is found in **isAvailable** for **FrameworkSingleBatteryBAT1** class.

So in class **FrameworkSingleBatteryBAT1** function **isAvailable**  you will have to add this.
```
        if (fileExists(BAT1_START_PATH))
            return false;
```

So it should look like this
```
   isAvailable() {
        if (!dmiVendor()?.includes('Framework'))
            return false;

        if (fileExists(BAT1_START_PATH))
            return false;

        this._supportedConfiguration = [];

        const hasSysfsNode = fileExists(BAT1_END_PATH);
        if (hasSysfsNode)
            this._supportedConfiguration.push('sysfs');

        this._frameworkToolPath = findValidProgramInPath('framework_tool');
        if (this._frameworkToolPath === null)
            this._frameworkToolPath = fileExists(FRAMEWORK_TOOL_PATH) ? FRAMEWORK_TOOL_PATH : null;
        if (this._frameworkToolPath)
            this._supportedConfiguration.push('framework-tool');

        if (this._supportedConfiguration.length <= 0)
            return false;

        this._settings.set_strv('multiple-configuration-supported', this._supportedConfiguration);
        if (this._supportedConfiguration.length === 1)
            this._settings.set_string('configuration-mode', this._supportedConfiguration[0]);
        return true;
    }
```

And also in the new class (copied from system76) rename everything from BAT0 to BAT1.

Sorry I am away from linux station, or would have done it myself.

**Comment by maniacx** [comment]
Look good me
Did you test it? Does it work? Also you seeing customization option for start threshold in preferences?

Also the end/start/diffierential limits does framework laptop have?
For example.
Dell laptops maximum start threshold it can set is 95% and not 98%.. 
Dell laptop the difference should be not less then 5%. So we cannot set  end=80% and start=78%. It should be end=80% and start=75%. Does frameworks have any limit like that? It it doesnot we can leave it as it..

I will merge this for now. Do let me know though if it works and all prefs are available?

**Comment by Guldoman** [comment]
> Look good me Did you test it? Does it work? Also you seeing customization option for start threshold in preferences?

Yeah everything seems to be working fine.
image

```
$ cat /sys/class/power_supply/BAT1/charge_control_{end,start}_threshold
80
75
```

Seems to work fine with any value:
```
$ cat /sys/class/power_supply/BAT1/charge_control_{end,start}_threshold
82
81
```

> Also the end/start/diffierential limits does framework laptop have? Does frameworks have any limit like that? It it doesnot we can leave it as it..

Only limitation seems to be that start must be strictly less than end.
I'll test more in the coming days and will open an issue/PR if I find any issue.

**Comment by maniacx** [comment]
@Guldoman 

Where can I find the link or discussion regarding this?

`probe_with_fwk_charge_control=Y`

I want to update the readme.

**Comment by Guldoman** [comment]
> Where can I find the link or discussion regarding this?

https://lore.kernel.org/lkml/20240630-cros_ec-charge-control-v5-5-8f649d018c52@weissschuh.net/

TL;DR: the EC is ChromeOS-based and it has both the base ChromeOS charging controls and Framework custom ones.
To be able to use the ChromeOS controls (the ones used in this PR), one must not use the Framework methods to set the limits (either via the UEFI, or other means like `framework_tool`). If those were used, just set them back to 100%.

Until Framework adopts the ChromeOS controls, setting the `probe_with_fwk_charge_control` module parameter is needed to allow the driver to work.

To set the module parameter I created a `/etc/modprobe.d/fw_use_cros_charging.conf` file with
```
options cros_charge_control probe_with_fwk_charge_control=Y
```
inside.

**Comment by TitouanReal** [comment]
It seems like the README is not updated yet to include this solution @maniacx

**Comment by maniacx** [comment]
I didn't quite understand the  procedure to get this enabled.  I couldn't find much information. 

If a  new linux user wants to enable this,  they might probably need detailed instruction to enable this, which i do not have , so I didn't want to post something that could be misleading.

 If one of you know step by step how to enable this, and post it here, I could add  it to readme file.

**Comment by TitouanReal** [comment]
After uninstalling [framework-laptop-kmod](https://github.com/DHowett/framework-laptop-kmod) and [framework-system](https://github.com/FrameworkComputer/framework-system), and creating the file `/etc/modprobe.d/fw_use_cros_charging.conf` containing:

```
options cros_charge_control probe_with_fwk_charge_control=Y
```

I thought I had it working. `charge_control_start_threshold` and `charge_control_end_threshold` are exposed in `/sys/class/power_supply/BAT1/` and contain the value I assigned from the extension. However the laptop stays at `charge_control_end_threshold` and does not start the discharge.

Maybe @Guldoman knows about reproducible steps to make this working.

**Comment by Guldoman** [comment]
Is the threshold in the bios set to 100?
Also try disabling the new battery extender functionality (if your BIOS supports it already).

**Comment by maniacx** [comment]
> However the laptop stays at charge_control_end_threshold and does not start the discharge.

Isn't that the correct way, though?

It should NOT force the battery to discharge quickly to start_threshold.

 My laptop is asus, and it has only end threshold, and the start threshold is automatically set to end threshold - 2 and it takes almost 5+ hours to discharge 1% of battery.

May be wait for a longer time.

Also , also on Chromebook, the author has removed start threshold on some models, as it was forcing discharge of the battery

https://github.com/torvalds/linux/commit/c28dc9fc24f5fa802d44ef7620a511035bdd803e

**Comment by TitouanReal** [comment]
For some reason I expected the battery to actively discharge. My mistake.

It seems like the battery passively discharges just as expected. It is very slow (as should be the case), making it hard to make sure it works, but it went down at least 1% while being plugged in.

However, turning the laptop off stops the passive discharge and the battery goes right back up to `charge_control_end_threshold`. Is it expected?

**Comment by Guldoman** [comment]
> However, turning the laptop off stops the passive discharge and the battery goes right back up to `charge_control_end_threshold`. Is it expected?

It's possible. I haven't looked at it too much so I can't say for sure, also because I usually just suspend it.
I'll keep an eye out the next few days for this behavior.

**Comment by Guldoman** [comment]
> Also , also on Chromebook, the author has removed start threshold on some models, as it was forcing discharge of the battery
> 
> [torvalds/linux@c28dc9f](https://github.com/torvalds/linux/commit/c28dc9fc24f5fa802d44ef7620a511035bdd803e)

It looks like this was backported to 6.12.8 and it affects my Framework laptop too, so the extension is falling back to the normal Framework sysfs method...

Still, documentation on how to enable the "native" sysfs method (without any external kernel module) might be useful.
The steps are:
A new(er) kernel is needed (6.11+).
1. Uninstall `framework-laptop-kmod` or any other similar kernel modules.
2. Reset the battery threshold in the BIOS back to 100%.
3. Run `sudo tee /etc/modprobe.d/fw_use_cros_charging.conf <<< "options cros_charge_control probe_with_fwk_charge_control=Y"`. This tells the driver that we want to use it, instead of the custom Framework implementation.
4. Reboot.
5. Check if the driver is enabled with `journalctl -b0 | grep "Framework charge control detected, preventing load"`. If this command has no output, the driver is likely working. Otherwise you might still have some Framework-specific threshold settings or kernel modules.

> [!NOTE]
> If you notice weird behavior after a few days (like the laptop charging above your set threshold), you might need to disable the battery extender functionality from the BIOS.
> Keep it enabled if you intend to use the laptop plugged in for an extended amount of time (even just a few days).
